create_asset_cache.go

```diff
@@ -39,20 +42,84 @@ func (c *Client) CreateAssetCache(ctx context.Context, opts CreateAssetCacheOpti
 	return assetCacheImage.Save()
 }
 
-// TODO -Dan- downloads should be concurrent.
+type downloadResult struct {
+	index  int
+	pair   dist.BlobAssetPair
+	sha256 string
+	err    error
+}
+
+type downloadJob struct {
+	asset dist.Asset
+	index int
+}
+
+// TODO -Dan- parallel downloads should cleanly exit with Ctrl-C
+// TODO -Dan- parallel download output is a bit messed up.
+// existing behavior is a bit dangerous and can poison the cache.
 func (c *Client) downloadAssets(assets []dist.Asset) (dist.BlobMap, error) {
-	result := make(dist.BlobMap)
-	for _, asset := range assets {
-		if asset.URI == "" {
-			continue
+	resultMap := make(dist.BlobMap)
+	results := make(chan downloadResult, len(assets))
+	jobs := make(chan downloadJob, len(assets))
+	for workerCount := 0; workerCount < assetDownloadWorkers; workerCount++ {
+		go downloadWorker(c.downloader, jobs, results)
+	}
+
+	for assetIdx, asset := range assets {
+		jobs <- downloadJob{
+			asset: asset,
+			index: assetIdx,
+		}
+	}
+	close(jobs)
+
+	resultIndiciesMap := map[string]int{}
+	errors := []error{}
+	for i := 0; i < len(assets); i++ {
+		r := <-results
+		prevIdx, ok := resultIndiciesMap[r.sha256]
+		switch {
+		case r.err != nil:
+			errors = append(errors, r.err)
+		case !ok || prevIdx < r.index:
```

Comment:
I'm struggling to see why we need `r.index` here. Our final return type is a mapping so preserving order doesn't matter.
Are we expecting duplicates to be returned from the channel with a given `r.sha256`?

Reply:
> Are we expecting duplicates to be returned from the channel with a given r.sha256?

Exactly. Two buildpacks share an asset, it has the same `r.sha256` but different metadata. If we don't always pick the same metadata we are missing a little bit of reproducibility (though this should be rare...).

If we remove this constraint and fiddle with the timings [here](https://github.com/buildpacks/pack/pull/1085/files#diff-27a7971b2d46c736fecf53ce767289b1b4121623503b68948654b51f8333a98aR96-R101), we can get failures.